### DISCUSSION TITLE
v4.0.0rc2 DB_HOST not working with ip addresses

### DISCUSSION THREAD

**Discussion by garyx** [mastodon/mastodon#20180] [Q&A]
I was thinking of creating a bug for this issue but first wanted to see if anyone here could help

1. I have setup v4.0.0rc2 using the install from source setup instructions. Everything works fine if I run postgresql locally and have **/var/run/postgresql** for DB_HOST
2. I wanted to move postgres to run separately on another VM. Copied db, accounts look good. pg_hba.conf in good shape. I can telnet from the mastodon VM, I can psql into the db, I even setup pgbouncer and that works fine manually and telnet to the postgresql port works fine. 
3. But no matter what I do in .env.production. Set it to localhost, ip address or no matter what I cannot get the mastodon-web and other services to connect. It just core-dumps and we see that it cannot connect. 
4. I even manually set the databases.yml and input a db url there. Still fails. 
5. And if I check the db logs it's like the mastodon services aren't even trying to connect, they seem to fail before 

Has anyone else tried v4.0.0rc2 and tried using a postgresql db through an ip address?

**Comment by MrGeneration** [comment]
That shouldn't be version specific but configuration specific.
If your `pga_hba.conf` does allow your Mastodon host by its IP (if you have IPv6 and IPv4 you may wanna double tap that) then ensure that the IP of your database server in the configuration is correct.

If all of above fits, check what exactly the error message is. Possibly the database server log helps as well.
Note that, especially if your database connection goes over the internet you may wanna use encryption in between these hosts (aka using SSL for your database connection).

**Comment by garyx** [comment]
Yeah the pga_hba.conf looks fine and as I said I can log in from the mastodon vm using other tools just fine, even manually with pgbouncer works to connect to the DB. 

This is a closed network and they are just adjacent to each other in a /24 subnet that I control. So nothing there is stopping it. 

The error message is that the database server refused connection, however the postgres logs have no attempts seen from mastodon. 

There is scram-sha-256 set to that connection in pga_hba.conf, but that shouldn't matter as far as I can tell. Mastadon should be able to use that

I set it up pgbouncer and just setting the vm to that localhost address didn't work either and pgabouncer did not see the connection come in from mastodon services. 

I spent 8 hours today testing various settings, tools and double checking the connection to the DB VM. Manually and with any other tool I could connect. But Mastodon crashes and it looks like it is crashing before it connects to anything.

**Comment by garyx** [comment]
The install procedure was to install V3.5.3. Then upgrade to RC1 and then RC2 of V4.0.0

**Comment by garyx** [comment]
I'll add more info here because I'm not sure what broke in the local setup. 

But I did fetch the repo again on another machine. Built everything with docker. Copied the build container over to the machine I had everything running on in the beginning. Used the exact same .env.production and voila, mastodon didn't crash anymore with an ip address in the DB_HOST parameter. 

I'm a bit lost what might have been the issue. But here we are and things work now.
